File: 剑指 Offer（第 2 版）/09.cpp
// 剑指 Offer 09. 用两个栈实现队列
// https://leetcode-cn.com/problems/yong-liang-ge-zhan-shi-xian-dui-lie-lcof/

/* 
维护两个栈，第一个栈支持插入操作，第二个栈支持删除操作。

插入 直接 push 到第一个栈里
删除 如果第二个栈为空，就把第一个栈的元素依次 pop 出来放到第二个栈里，从第二个栈弹出
    如果第二个栈不为空，直接弹出栈顶元素 
*/
#include <stack>
using namespace std;

class CQueue {
    stack<int> stack1,stack2;
public:
    CQueue() {
        while (!stack1.empty()) {
            stack1.pop();
        }
        while (!stack2.empty()) {
            stack2.pop();
        }
    }
    
    void appendTail(int value) {
        stack1.push(value);
    }
    
    int deleteHead() {
        if (stack2.empty()) {
            while (!stack1.empty()) {
                stack2.push(stack1.top());
                stack1.pop();
            }
        } 
        if (stack2.empty()) {
            return -1;
        } else {
            int deleteItem = stack2.top();
            stack2.pop();
            return deleteItem;
        }
    }
};

File: 剑指 Offer（第 2 版）/09.java
import java.util.LinkedList;

class CQueue {
    LinkedList<Integer> stack1;
    LinkedList<Integer> stack2;
    public CQueue() {
        stack1 = new LinkedList<>();
        stack2 = new LinkedList<>();
    }
    
    public void appendTail(int value) {
        stack1.add(value);
    }
    
    public int deleteHead() {
        if(stack2.isEmpty()){
            while(!stack1.isEmpty()){
                stack2.add(stack1.removeLast());
            }
        }
        if(stack2.isEmpty()){
            return -1;
        }
        int ans = stack2.removeLast();
        return ans;
    }
}